app/upgrade_cudos.go

```diff
@@ -585,76 +592,121 @@ func parseGenesisAccounts(jsonData map[string]interface{}, contractAccountMap *O
+				resolvedDelegatorMap.Set(validatorOperatorAddress, resolvedDelegator.Add(delegatorTokens))
+				unbondingDelegatedBalanceMap.Set(resolvedDelegatorAddress, resolvedDelegatorMap)
+			}
+		}
+	}
+
+	return delegatedBalanceMap, unbondingDelegatedBalanceMap, nil
+}
+
+func parseGenesisUnbondingDelegations(validators *OrderedMap[string, *ValidatorInfo], contracts *OrderedMap[string, *ContractInfo], cudosCfg *CudosMergeConfig) (*OrderedMap[string, *OrderedMap[string, sdk.Int]], error) {
+	// Handle delegations
+	unbondedDelegatedBalanceMap := NewOrderedMap[string, *OrderedMap[string, sdk.Int]]()
+
+	for i := range validators.Iterate() {
+		validatorOperatorAddress, validator := i.Key, i.Value
+
+		for j := range validator.UnbondingDelegations.Iterate() {
+			delegatorAddress, delegation := j.Key, j.Value
+
+			resolvedDelegatorAddress, err := resolveIfContractAddressWithFallback(delegatorAddress, contracts, cudosCfg)
+			if err != nil {
+				return nil, err
 			}
+
+			delegatorTokens := sdk.NewInt(0)
+
+			for _, entry := range delegation.Entries {
+				delegatorTokens = delegatorTokens.Add(entry.Balance)
+			}
+
+			if delegatorTokens.IsZero() {
+				// This happens when number of shares is less than 1
+				continue
+			}
+
+			// Store delegation to delegated map
+			resolvedDelegatorMap, _ := unbondedDelegatedBalanceMap.GetOrSetDefault(resolvedDelegatorAddress, NewOrderedMap[string, sdk.Int]())
+			resolvedDelegator, _ := resolvedDelegatorMap.GetOrSetDefault(validatorOperatorAddress, sdk.NewInt(0))
+			resolvedDelegatorMap.Set(validatorOperatorAddress, resolvedDelegator.Add(delegatorTokens))
+			unbondedDelegatedBalanceMap.Set(resolvedDelegatorAddress, resolvedDelegatorMap)
 		}
 	}
 
-	return delegatedBalanceMap, nil
+	return unbondedDelegatedBalanceMap, nil
 }
 
 type DelegationInfo struct {
-	delegatorAddress string
-	shares           sdk.Dec
+	DelegatorAddress string
+	Shares           sdk.Dec
 }
 
 type UnbondingDelegationInfo struct {
-	delegatorAddress string
-	entries          []*UnbondingDelegationEntry
+	DelegatorAddress string
+	Entries          []*UnbondingDelegationEntry
 }
 
 type UnbondingDelegationEntry struct {
-	balance        sdk.Int
-	initialBalance sdk.Int
-	creationHeight uint64
-	completionTime string
+	Balance        sdk.Int
+	InitialBalance sdk.Int
+	CreationHeight uint64
+	CompletionTime string
 }
 
 type ValidatorInfo struct {
-	stake                sdk.Int
-	shares               sdk.Dec
-	status               string
-	operatorAddress      string
-	consensusPubkey      cryptotypes.PubKey
-	delegations          *OrderedMap[string, *DelegationInfo]
-	unbondingDelegations *OrderedMap[string, *UnbondingDelegationInfo]
+	Stake                sdk.Int
+	Shares               sdk.Dec
+	Status               string
+	OperatorAddress      string
+	ConsensusPubkey      cryptotypes.PubKey
+	Delegations          *OrderedMap[string, *DelegationInfo]
+	UnbondingDelegations *OrderedMap[string, *UnbondingDelegationInfo]
```

Comment:
It feels like that `ValidatorInfo` struct should  contain also the `UndondedDelegations` and `Rewards` maps.

Reply:
This map mimics the structure of genesis json.

I've moved rewards to distibution info where they are calculated.